Support NXP LPADC and regulator(VREF) in MCXN

ZhaoxiangJin opened the pull request:
Support NXP LPADC and regulator(VREF) in MCXN.
The VREF provides the bias current for LPADC, and the VREF output could be selected as LPADC reference.
In this PR, the LPADC selects the VREF output as the reference, and the VREF output is 1.8v. The user could use the new property **nxp,reference-supply-value** of LPADC to set the VREF output.

danieldegrasse commented on boards/nxp/frdm_mcxn947/board.c:
```diff
@@ -6,6 +6,9 @@
 #include <zephyr/device.h>
 #include <fsl_clock.h>
 #include <fsl_spc.h>
+#if defined(CONFIG_SOC_MCXN947) && defined(CONFIG_ADC_MCUX_LPADC)
```
Only the `CONFIG_ADC_MCUX_LPADC` needs to be present here- this file is specific to the MCXN947 SOC

ZhaoxiangJin replied:
> Only the `CONFIG_ADC_MCUX_LPADC` needs to be present here- this file is specific to the MCXN947 SOC

Yes, this is mcxn947 specific file. I will update it.

decsny commented on boards/nxp/frdm_mcxn947/board.c:
```diff
@@ -7,6 +7,9 @@
 #include <fsl_spc.h>
 #include <soc.h>
+#if defined(CONFIG_ADC_MCUX_LPADC)
+#include <fsl_vref.h>
```
is this needed? I don't see anything from fsl_vref.h being used in this file

ZhaoxiangJin replied:
removed

decsny reviewed (changes requested):
Don't understand why pulling in the vref driver in the build, it doesn't seem used or necessary

ZhaoxiangJin commented:
Hi @decsny, I'd like to have your comments here, thanks.

Albort12138 commented on dts/arm/nxp/nxp_mcxn94x_common.dtsi:
```diff
@@ -585,6 +585,15 @@
+
+	vref: vref@111000 {
+		compatible = "nxp,vref";
+		regulator-name = "lpc55s36-vref";
```
why use this name? This update is for mcxn94x right?

ZhaoxiangJin replied:
same ip, lpc55s36-vref is applicable for the n10.

danieldegrasse replied:
Sure, but the `compatible` string manages the driver used. The `regulator-name` property is documented as "A string used as a descriptive name for regulator outputs". I think we can change this to "mcxn94x-vref" for this SOC.

ZhaoxiangJin replied:
Yes, which driver to use is determined by the compatible property, the regulator-name property is not for this.

decsny reviewed (changes requested):
tests/drivers/regulator/api does not build

decsny commented:
```
west build -b frdm_mcxn947//cpu0 -p always tests/drivers/regulator/api/
-- west build: making build dir /home/declan/zephyrproject/zephyr/build pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /home/declan/zephyrproject/zephyr/tests/drivers/regulator/api
-- CMake version: 3.28.3
-- Found Python3: /home/declan/zephyrproject/.venv/bin/python3 (found suitable version "3.10.12", minimum required is "3.8") found components: Interpreter
-- Cache files will be written to: /home/declan/.cache/zephyr
-- Zephyr version: 3.6.99 (/home/declan/zephyrproject/zephyr)
-- Found west (found suitable version "1.1.0", minimum required is "0.14.0")
-- Board: frdm_mcxn947, qualifiers: mcxn947/cpu0
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.16.3 (/home/declan/zephyr-sdk-0.16.3)
-- Found toolchain: zephyr 0.16.3 (/home/declan/zephyr-sdk-0.16.3)
-- Found Dtc: /usr/bin/dtc (found suitable version "1.6.1", minimum required is "1.4.6")
-- Found BOARD.dts: /home/declan/zephyrproject/zephyr/boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0.dts
-- Found devicetree overlay: /home/declan/zephyrproject/zephyr/tests/drivers/regulator/api/app.overlay
-- Generated zephyr.dts: /home/declan/zephyrproject/zephyr/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: /home/declan/zephyrproject/zephyr/build/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: /home/declan/zephyrproject/zephyr/build/zephyr/dts.cmake
Parsing /home/declan/zephyrproject/zephyr/Kconfig
Loaded configuration '/home/declan/zephyrproject/zephyr/boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0_defconfig'
Merged configuration '/home/declan/zephyrproject/zephyr/tests/drivers/regulator/api/prj.conf'
Configuration saved to '/home/declan/zephyrproject/zephyr/build/zephyr/.config'
Kconfig header saved to '/home/declan/zephyrproject/zephyr/build/zephyr/include/generated/autoconf.h'
-- Found GnuLd: /home/declan/zephyr-sdk-0.16.3/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd (found version "2.38")
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /home/declan/zephyr-sdk-0.16.3/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
Load components for MCXN947_cm33_core0:
driver_common component is included.
driver_reset component is included.
device_CMSIS component is included.
CMSIS_Include_core_cm component is included.
device_system component is included.
driver_ostimer component is included.
driver_lpflexcomm component is included.
driver_lpuart component is included.
driver_mcx_spc component is included.
driver_cache_cache64 component is included.
driver_port component is included.
-- Using ccache: /usr/bin/ccache
-- Configuring done (6.3s)
-- Generating done (0.1s)
-- Build files have been written to: /home/declan/zephyrproject/zephyr/build
-- west build: building application
[1/145] Preparing syscall dependency handling

[2/145] Generating include/generated/version.h
-- Zephyr version: 3.6.99 (/home/declan/zephyrproject/zephyr), build: v3.6.0-1891-gf463c02ee3b7
[129/145] Building C object zephyr/drivers/regul.../drivers__regulator.dir/regulator_nxp_vref.c.obj
FAILED: zephyr/drivers/regulator/CMakeFiles/drivers__regulator.dir/regulator_nxp_vref.c.obj
ccache /home/declan/zephyr-sdk-0.16.3/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DCPU_MCXN947VDF_cm33_core0 -DKERNEL -DK_HEAP_MEM_POOL_SIZE=0 -DLPFLEXCOMM_INIT_NOT_USED_IN_DRIVER=1 -DPICOLIBC_LONG_LONG_PRINTF_SCANF -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/declan/zephyrproject/zephyr/include -I/home/declan/zephyrproject/zephyr/build/zephyr/include/generated -I/home/declan/zephyrproject/zephyr/soc/nxp/mcx -I/home/declan/zephyrproject/zephyr/soc/nxp/mcx/mcxnx4x/. -I/home/declan/zephyrproject/zephyr/subsys/testsuite/include -I/home/declan/zephyrproject/zephyr/subsys/testsuite/ztest/include -I/home/declan/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include -I/home/declan/zephyrproject/zephyr/modules/cmsis/. -I/home/declan/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/devices/MCXN947 -I/home/declan/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/devices/MCXN947/drivers -I/home/declan/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/common -I/home/declan/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/ostimer -I/home/declan/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/lpflexcomm -I/home/declan/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/lpflexcomm/lpuart -I/home/declan/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/mcx_spc -I/home/declan/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/cache/cache64 -I/home/declan/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/mcx_romapi/flash -I/home/declan/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/drivers/port -I/home/declan/zephyrproject/zephyr/modules/hal_nxp/. -isystem /home/declan/zephyrproject/zephyr/lib/libc/common/include -Wshadow -fno-strict-aliasing -Os -imacros /home/declan/zephyrproject/zephyr/build/zephyr/include/generated/autoconf.h -fno-printf-return-value -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m33 -mthumb -mabi=aapcs -mfp16-format=ieee -mtp=soft --sysroot=/home/declan/zephyr-sdk-0.16.3/arm-zephyr-eabi/arm-zephyr-eabi -imacros /home/declan/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -ftls-model=local-exec -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home/declan/zephyrproject/zephyr/tests/drivers/regulator/api=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/declan/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/declan/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -mcmse --specs=picolibc.specs -std=c99 -MD -MT zephyr/drivers/regulator/CMakeFiles/drivers__regulator.dir/regulator_nxp_vref.c.obj -MF zephyr/drivers/regulator/CMakeFiles/drivers__regulator.dir/regulator_nxp_vref.c.obj.d -o zephyr/drivers/regulator/CMakeFiles/drivers__regulator.dir/regulator_nxp_vref.c.obj -c /home/declan/zephyrproject/zephyr/drivers/regulator/regulator_nxp_vref.c
/home/declan/zephyrproject/zephyr/drivers/regulator/regulator_nxp_vref.c: In function 'regulator_nxp_vref_init':
/home/declan/zephyrproject/zephyr/drivers/regulator/regulator_nxp_vref.c:193:23: error: 'VREF_CSR_REFL_GRD_SEL_MASK' undeclared (first use in this function); did you mean 'VREF_CSR_REGEN_MASK'?
  193 |         base->CSR &= ~VREF_CSR_REFL_GRD_SEL_MASK;
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                       VREF_CSR_REGEN_MASK
/home/declan/zephyrproject/zephyr/drivers/regulator/regulator_nxp_vref.c:193:23: note: each undeclared identifier is reported only once for each function it appears in
[137/145] Building C object zephyr/kernel/CMakeFiles/kernel.dir/timeout.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/declan/zephyrproject/zephyr/build
```

decsny commented:
the driver does not appear to build with mcx for me, did it build when you tested it?

mmahadevan108 commented on drivers/regulator/regulator_nxp_vref.c:
```diff
@@ -183,15 +183,19 @@ static const struct regulator_driver_api api = {
 
 	regulator_common_data_init(dev);
 
+#if !(defined CONFIG_SOC_MCXN947)
```
This should be captured in Kconfig instead of a platform define.

mmahadevan108 commented:
@ZhaoxiangJin, how was this driver tested?

ZhaoxiangJin commented:
> @ZhaoxiangJin, how was this driver tested?

I tested the driver API, the function test is blocked by the LPADC enablement

mmahadevan108 commented:
@ZhaoxiangJin, kindly resolve the conflicts.

ZhaoxiangJin committed:
dts: arm/nxp/mcxn94x: Add lpadc nodes for NXP mcxn94x

Add lpadc nodes for NXP mcxn94x

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: nxp/frdm_mcxn947: Support lpadc for NXP frdm_mcxn947

Support lpadc for NXP frdm_mcxn947

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
samples: drivers/adc: Support adc example for NXP frdm_mcxn947

Support adc example for NXP frdm_mcxn947

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
tests: drivers/adc/adc_api: Support adc api test for NXP frdm_mcxn947

Support adc api test for NXP frdm_mcxn947

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

mmahadevan108 commented on drivers/regulator/regulator_nxp_vref.c:
```diff
@@ -183,15 +183,21 @@ static const struct regulator_driver_api api = {
 	base->CSR &= ~VREF_CSR_REFL_GRD_SEL_MASK;
 	base->CSR |= config->gnd_sel;
+#endif /* CONFIG_REGULATOR_NXP_VREF_HAS_REFL_GRD_SEL */
 
```
Please delete this block of code and document it in the commit.

mmahadevan108 commented on drivers/regulator/regulator_nxp_vref.c:
```diff
@@ -183,15 +183,21 @@ static const struct regulator_driver_api api = {
 {
+#if (defined(CONFIG_REGULATOR_NXP_VREF_HAS_REFL_GRD_SEL) \
+	&& CONFIG_REGULATOR_NXP_VREF_HAS_REFL_GRD_SEL)
 	const struct regulator_nxp_vref_config *config = dev->config;
```
Please delete this block of code and document it in the commit message.

mmahadevan108 commented on drivers/regulator/Kconfig.nxp_vref:
```diff
@@ -12,3 +12,10 @@ config REGULATOR_NXP_VREF_INIT_PRIORITY
+	default n if SOC_MCXN947
+	default y if SOC_LPC55S36
+	help
+	  Decides whether NXP VREF peripheral has ground select.
```
No need for this as the suggestion from the VREF designer is to delete the use of `CSR_REFL_GRL_SEL` bit

ZhaoxiangJin committed:
dts: arm/nxp/mcxn94x: Add vref node for NXP MCXN94x

Add the vref node for NXP MCXN94x

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: nxp/frdm_mcxn947: Support vref for NXP frdm_mcxn947

Support vref for NXP frdm_mcxn947

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

danieldegrasse commented on dts/bindings/regulator/nxp,vref.yaml:
```diff
@@ -20,12 +20,6 @@ properties:
   reg:
     required: true
 
-  nxp,ground-select:
```
This commit needs to be after the change dropping the `gnd_sel` field from the driver

ZhaoxiangJin replied:
yes

danieldegrasse reviewed (commented):
I'm a bit unclear why we are dropping support for the `GND_SEL` field? This is documented in both the MCX and LPC55S36 reference manuals, and I don't see anything claiming this cannot be used by the user. Perhaps we should document this restriction in the binding YAML? My concern is given that this bit is documented, someone could reenable support for it in the future.

ZhaoxiangJin commented:
> I'm a bit unclear why we are dropping support for the `GND_SEL` field? This is documented in both the MCX and LPC55S36 reference manuals, and I don't see anything claiming this cannot be used by the user. Perhaps we should document this restriction in the binding YAML? My concern is given that this bit is documented, someone could reenable support for it in the future.

Hi Daniel, I believe there is some information misaligned between us, I forwarded an email to you, please kindly receive it.

danieldegrasse commented:
> Hi Daniel, I believe there is some information misaligned between us, I forwarded an email to you, please kindly receive it.

Thank you for the information- given this, I am ok to deprecate this field from the driver. We may want to add a note within the migration guide for Zephyr 3.7, so that users are aware this property was removed and is not supported. We can add this as an additional commit within this PR, to the `migration-guide-3.7.rst` file

ZhaoxiangJin committed:
docs: releases/migration-guide-3.7: Add NXP regulator driver info

Add NXP regulator (nxp,vref) driver info

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

danieldegrasse commented on drivers/regulator/regulator_nxp_vref.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2023 NXP
+ * Copyright 2024 NXP
```
Copyright year here should not be replaced, instead convert it to a range
```suggestion
 * Copyright 2023-2024 NXP
```

danieldegrasse commented on dts/bindings/regulator/nxp,vref.yaml:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2023 NXP
+# Copyright 2024 NXP
```
```suggestion
# Copyright 2023-2024 NXP
```

decsny reviewed (changes requested):
still confused how this was tested since none of the regulator hardware tests were enabled and in fact dont build

(there was some confusion about the api test, it only tests the fake regulator)

ZhaoxiangJin commented:
> still confused how this was tested since none of the regulator hardware tests were enabled and in fact dont build
> 
> (there was some confusion about the api test, it only tests the fake regulator)

Fine, I will combine LPADC VREF enablement PRs, use the VREF output voltage as the LPADC reference voltage.

ZhaoxiangJin committed:
drivers: adc: Support regulator output as reference to LPADC

Support regulator output different voltage as reference to LPADC
the regulator output voltage can be set through property
zephyr,vref-mv

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

decsny commented:
I still think it would be good to enable the VREF on MCX on one of the regulator tests

ZhaoxiangJin committed:
tests: drivers/regulator/voltage: Add regulator test for NXP frdm_mcxn947

Support regulator(vref) output vt oltage test for NXP frdm_mcxn947

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented:
> I still think it would be good to enable the VREF on MCX on one of the regulator tests

Added regulator(vref) test case. please review, thanks.

danieldegrasse commented on drivers/regulator/regulator_nxp_vref.c:
```diff
@@ -165,7 +164,7 @@ static int regulator_nxp_vref_get_voltage(const struct device *dev,
 	idx = (base->UTRIM & VREF_UTRIM_TRIM2V1_MASK) >> VREF_UTRIM_TRIM2V1_SHIFT;
 
-	ret = linear_range_get_value(&utrim_range, base->UTRIM, volt_uv);
+	ret = linear_range_get_value(&utrim_range, idx, volt_uv);
```
This looks like a separate fix from the `gnd_sel` issue- maybe this line needs to be in another commit? This will make it easier for users in the future to follow *why* this change was made

ZhaoxiangJin replied:
yes, this is bug fix, I can add one commit to do this

danieldegrasse commented on dts/bindings/adc/nxp,lpc-lpadc.yaml:
```diff
@@ -75,6 +75,10 @@ properties:
     type: phandles
     description: References to required regulators which must be enabled for LPADC to function
 
+  nxp,reference-supply-value:
```
I don't think we need to add this property- instead, we can read the `zephyr,vref-mv` property, and attempt to set the regulator to that voltage

ZhaoxiangJin replied:
yes, the property zephyr,vref-mv can do this

danieldegrasse commented on drivers/adc/adc_mcux_lpadc.c:
```diff
@@ -552,6 +570,7 @@ static const struct adc_driver_api mcux_lpadc_driver_api = {
 		.irq_config_func = mcux_lpadc_config_func_##n,				\
 		.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),			\
 		.ref_supplies = mcux_lpadc_ref_supplies_##n, \
+		.ref_voltage_val = DT_INST_PROP_OR(n, nxp_reference_supply_value, 0), \
```
See comment on DT binding- I think we should extract this value from `zephyr,vref-mv`

danieldegrasse commented on tests/drivers/regulator/voltage/boards/frdm_mcxn947_mcxn947_cpu0.overlay:
```diff
@@ -0,0 +1,44 @@
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	voltage-ref= <2>;
```
Just a nit- might be worth mentioning here in a comment that this switches the ADC source to no longer use the VREF as a source.

Also, won't the LPADC driver still try to set the regulator's voltage in this case? Perhaps we need to delete this property from the node for this overlay with the `/delete-node/` property

ZhaoxiangJin replied:
the LPADC will not set the regulator's voltage, in this case, the **zephyr,reference** property is set to **ADC_REF_EXTERNAL0**, which means the regulator will not be stetted in the LPADC driver.

ZhaoxiangJin committed:
drivers: regulator/nxp_vref: Fixed incorrect input parameter

Fixed incorrect parameter passed in function linear_range_get_value.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: regulator/nxp_vref: Remove nxp_ground_select property

Remove nxp_ground_select property.
Delete the use of NXP vref peripheral CSR register REFL_GRL_SEL bit.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

danieldegrasse commented on drivers/adc/adc_mcux_lpadc.c:
```diff
@@ -200,8 +205,22 @@ static int mcux_lpadc_channel_setup(const struct device *dev,
+	if (channel_cfg->reference == ADC_REF_EXTERNAL1) {
+		LOG_DBG("ref external1");
+		if (*regulator != NULL) {
+			regulator_set_voltage(*regulator, min_uv, max_uv);
```
Thanks for making this update- we might want to check the return code here, and propagate it to the caller of this function if the call fails:

```
ret = regulator_set_voltage(...);
if (ret < 0) {
    return ret;
}
```

dleach02 commented:
@decsny can you clarify what your specific change request is?

decsny commented:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/110566814/8f68a369-ed7f-4682-bfff-fc0a4b824f8a)

looks good now

decsny commented on boards/nxp/frdm_mcxn947/doc/index.rst:
```diff
@@ -82,6 +82,10 @@ The FRDM-MCXN947 board configuration supports the following hardware features:
 +-----------+------------+-------------------------------------+
 | CTIMER    | on-chip    | counter                             |
 +-----------+------------+-------------------------------------+
+| REGULATOR | on-chip    | vref                                |
```
should be VREF on the left column and regulator on the right column

decsny commented on dts/bindings/regulator/nxp,vref.yaml:
```diff
@@ -20,12 +20,6 @@ properties:
   reg:
     required: true
 
-  nxp,ground-select:
```
please squash the binding change to the commit with the driver change

decsny commented on drivers/adc/adc_mcux_lpadc.c:
```diff
@@ -122,10 +122,16 @@ static int mcux_lpadc_acquisition_time_setup(const struct device *dev, uint16_t
+	const struct device **regulator = config->ref_supplies;
+	uint16_t vref_mv = CONTAINER_OF(channel_cfg, struct adc_dt_spec, channel_cfg)->vref_mv;
+	int32_t min_uv = (int32_t)((uint32_t)vref_mv * 1000);
+	int32_t max_uv = (int32_t)((uint32_t)vref_mv * 1000);
```
why have two variables if they are the same?

ZhaoxiangJin replied:
Considering saving stack, using one variable in the current situation is ok, I will update it.

decsny commented on drivers/adc/adc_mcux_lpadc.c:
```diff
@@ -200,8 +206,25 @@ static int mcux_lpadc_channel_setup(const struct device *dev,
+	 * ADC_REF_EXTERNAL0: Use other voltage source (maybe also within the SoCs)
+	 * as LPADC reference voltage, like VREFH, VDDA, etc.
+	 */
+	if (channel_cfg->reference == ADC_REF_EXTERNAL1) {
```
do we need to add more options to the ADC ref enum in the api header?

ZhaoxiangJin replied:
Maybe not, and I would like to change the **ADC_REF_EXTERNAL1** to **ADC_REF_INTERNAL** and add **ref_internal** for the LPADC driver.

decsny reviewed (changes requested):
the PR is pretty good, the change request is just for the doc, other comments are optional

decsny commented on boards/nxp/frdm_mcxn947/doc/index.rst:
```diff
@@ -84,6 +84,10 @@ The FRDM-MCXN947 board configuration supports the following hardware features:
 +-----------+------------+-------------------------------------+
 | USDHC     | on-chip    | sdhc                                |
 +-----------+------------+-------------------------------------+
+| VREF      | on-chip    | REGULATOR                           |
```
nit: in case you have to push again for some reason, change regulator to lowercase please
